### PR TITLE
Remove InternalsVisibleTo from System.Collections.Immutable.

### LINKED ISSUE #107848: Remove IVT from System.Collections.Immutable
System.Collections.Immutable currently has InternalsVisibleTo to its tests. This is something we try hard to avoid. We should remove the reliance on internals from the tests and switch the tests over to referencing the reference assembly. Some of the validation can be moved to being debug-only asserts in the actual implementation. Other use of internals are purely for convenience. Other uses could be achieved via either reflection or more preferably by separating out the small amount of internals use into a separate unit tests project that builds the relevant product src into the test, ala what's done in networking tests, in regex, etc.

Fixing this will shrink the size of the product assembly (the linker will be able to trim it), will avoid future issues like https://github.com/dotnet/runtime/issues/107840, will enable our analyzers that bail on IVT to work for S.C.I, and will make it more consistent with the rest of dotnet/runtime.

### PR DESCRIPTION
**Pull request by eiriktsarpalis**
Fix #107848.

### CODE REVIEW

File: src/libraries/System.Collections.Immutable/tests/System.Collections.Immutable.Tests.csproj [hunk 1]
```diff
@@ -3,13 +3,20 @@
...
     <RdXmlFile Include="default.rd.xml" />
   </ItemGroup>
   <ItemGroup>
     <Compile Include="$(CommonTestPath)System\Collections\DictionaryExtensions.cs" Condition="'$(TargetFrameworkIdentifier)' == '.NETFramework'" Link="Common\System\Collections\DictionaryExtensions.cs" />
+    <!-- Reference internal product components for testing -->
+    <Compile Include="..\src\System\Polyfills.cs" />
+    <Compile Include="..\src\System\Collections\Frozen\String\Hashing.cs" Link="System\Collections\Frozen\String\Hashing.cs" />
+    <Compile Include="..\src\System\Collections\Frozen\String\KeyAnalyzer.cs" Link="System\Collections\Frozen\String\KeyAnalyzer.cs" />
+    <Compile Include="..\src\System\Collections\Immutable\IBinaryTree.cs" Link="System\Collections\Immutable\IBinaryTree.cs" />
+    <Compile Include="..\src\Validation\Requires.cs" Link="Validation\Requires.cs" />
```
Comment (stephentoub) [inline src/libraries/System.Collections.Immutable/tests/System.Collections.Immutable.Tests.csproj:19]: We should probably follow up to separate these out further, but for now this PR is definitely a good step in the right direction.

**Review by stephentoub — APPROVED** [review]
Thanks!

**Comment by stephentoub** [comment]
Out of curiosity, I tried building S.C.I with and without this PR. This reduces its size by ~5%.

File: src/libraries/System.Collections.Immutable/src/System.Collections.Immutable.csproj [hunk 2]
```diff
@@ -179,4 +176,8 @@ The System.Collections.Immutable library is built-in as part of the shared frame
     <PackageReference Include="System.Runtime.CompilerServices.Unsafe" Version="$(SystemRuntimeCompilerServicesUnsafeVersion)" />
   </ItemGroup>
 
+  <ItemGroup>
+    <Folder Include="System\Collections\Generic\" />
+  </ItemGroup>
```
Comment (stephentoub) [inline src/libraries/System.Collections.Immutable/src/System.Collections.Immutable.csproj]: Why is this needed?

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Collections.Immutable/src/System.Collections.Immutable.csproj (hunk 2 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Collections.Immutable/src/System.Collections.Immutable.csproj]: VS artifact, removed.

**Comment by eiriktsarpalis** [comment]
@vitek-karas could you suggest any workarounds for enabling private field access in blazor-wasm? My change in https://github.com/dotnet/runtime/pull/107872/commits/d3cf8f9cfd068204ab7ecc739f282db7e2e58b7c doesn't seem to do it.
